internal/shared/git/git.go

```diff
@@ -0,0 +1,237 @@
+// if that cannot be found, it assumes rawurl is a local path. If none
+// of these rules apply, Parse returns an error.
+func Parse(rawurl string) (u *url.URL, err error) {
+	parsers := []Parser{
+		ParseTransport,
+		ParseScp,
+		ParseLocal,
+	}
+
+	// Apply each parser in turn; if the parser succeeds, accept its
+	// result and return.
+	for _, p := range parsers {
+		u, err = p(rawurl)
+		if err == nil {
+			return u, err
+		}
+	}
+
+	// It's unlikely that none of the parsers will succeed, since
+	// ParseLocal is very forgiving.
+	return new(url.URL), fmt.Errorf("failed to parse %q", rawurl)
+}
+
+// ParseTransport parses rawurl into a URL object. Unless the URL's
+// scheme is a known Git transport, ParseTransport returns an error.
+func ParseTransport(rawurl string) (*url.URL, error) {
+	u, err := url.Parse(rawurl)
+	if err == nil && !Transports.Valid(u.Scheme) {
+		err = fmt.Errorf("scheme %q is not a valid transport", u.Scheme)
+	}
+	return u, err
+}
+
+// ParseScp parses rawurl into a URL object. The rawurl must be
+// an SCP-like URL, otherwise ParseScp returns an error.
+func ParseScp(rawurl string) (*url.URL, error) {
+	match := scpSyntax.FindAllStringSubmatch(rawurl, -1)
+	if len(match) == 0 {
+		return nil, fmt.Errorf("no scp URL found in %q", rawurl)
+	}
+	m := match[0]
+	user := strings.TrimRight(m[1], "@")
+	var userinfo *url.Userinfo
+	if user != "" {
+		userinfo = url.User(user)
+	}
+	rawquery := ""
+	if len(m) > 3 {
+		rawquery = m[4]
+	}
+	return &url.URL{
+		Scheme:   "ssh",
+		User:     userinfo,
+		Host:     m[2],
+		Path:     m[3],
+		RawQuery: rawquery,
+	}, nil
+}
+
+// ParseLocal parses rawurl into a URL object with a "file"
+// scheme. This will effectively never return an error.
+func ParseLocal(rawurl string) (*url.URL, error) {
+	return &url.URL{
+		Scheme: "file",
+		Host:   "",
+		Path:   rawurl,
+	}, nil
+}
+
+// GetRepoCloneOptions creates git.CloneOptions for cloning a repo with HTTP, SSH, or TLS credentials from Kubernetes secrets.
+func GetRepoCloneOptions(ctx context.Context, repoCred *controllerconfig.RepoCredential, kubeClient kubernetes.Client, namespace string, repo *v1alpha1.RepositoryPath) (*git.CloneOptions, error) {
+	endpoint, err := transport.NewEndpoint(repo.RepoUrl)
+	if err != nil {
+		return nil, err
+	}
+	var cloneOptions *git.CloneOptions
+	switch {
+	case repoCred.TLS != nil && repoCred.HTTPCredential != nil:
+		cred := repoCred.HTTPCredential
+		username := cred.Username
+		secret, err := kubeClient.GetSecret(ctx, namespace, cred.Password.Name)
+		if err != nil {
+			return nil, err
+		}
+		password := string(secret.Data[cred.Password.Key])
+		authMethod := &gitHttp.BasicAuth{
+			Username: username,
+			Password: password,
+		}
+		cloneOptions = &git.CloneOptions{
+			URL:             endpoint.String(),
+			Auth:            authMethod,
+			InsecureSkipTLS: repoCred.TLS.InsecureSkipVerify,
+		}
+		return cloneOptions, nil
```

Comment:
I believe TLS can be combined with SSH. I think TLS is an orthogonal concept to HTTP vs SSH. So, I am thinking if you can essentially remove the first case statement, and have the latter 2 set `cloneOptions`. Then you can take what you have, and if TLS is also set then set TLS settings in addition.

Reply:
Oh wait. Also, we need to be using the protocol defined by the RepositoryPath. Here's what I imagine in the ConfigMap:
```
repoCredentials:
  - url: "github.com/numaproj-labs"
    httpCredential:
      ...
    sshCredential:
      ...
    tlsCredential:
      ...
```
(Note that all of those sections are optional.)
When we clone, our RepositoryPath will be something like "https://github.com/numaproj/numaflow.git". We need to take that RepositoryPath and say "okay, it starts with https", so let me only look to see whether there's an HTTPCredential. Then, in addition I can look for a TLSCredential.

Reply:
I have made the method switch as per the scheme of git url